include/nbl/builtin/hlsl/matrix_utils/transformation_matrix_utils.hlsl

```diff
@@ -0,0 +1,112 @@
+#ifndef _NBL_BUILTIN_HLSL_TRANSFORMATION_MATRIX_UTILS_INCLUDED_
+#define _NBL_BUILTIN_HLSL_TRANSFORMATION_MATRIX_UTILS_INCLUDED_
+
+#include <nbl/builtin/hlsl/cpp_compat.hlsl>
+
+namespace nbl
+{
+namespace hlsl
+{
+
+template<typename T>
+matrix<T, 3, 4> getMatrix3x4As4x4(const matrix<T, 3, 4>& mat)
+{
+	matrix<T, 4, 4> output;
+	for (int i = 0; i < 3; ++i)
+		output[i] = mat[i];
+	output[3] = float32_t4(0.0f, 0.0f, 0.0f, 1.0f);
+
+	return output;
+}
+
+// TODO: use portable_float when merged
+//! multiplies matrices a and b, 3x4 matrices are treated as 4x4 matrices with 4th row set to (0, 0, 0 ,1)
+template<typename T>
+inline matrix<T, 3, 4> concatenateBFollowedByA(const matrix<T, 3, 4>& a, const matrix<T, 3, 4>& b)
+{
+	const matrix<T, 4, 4> a4x4 = getMatrix3x4As4x4(a);
+	const matrix<T, 4, 4> b4x4 = getMatrix3x4As4x4(b);
+	return matrix<T, 3, 4>(mul(a4x4, b4x4));
+}
+
+template<typename T>
+inline matrix<T, 3, 4> buildCameraLookAtMatrixLH(
+	const vector<T, 3>& position,
+	const vector<T, 3>& target,
+	const vector<T, 3>& upVector)
+{
+	const vector<T, 3> zaxis = core::normalize(target - position);
+	const vector<T, 3> xaxis = core::normalize(core::cross(upVector, zaxis));
+	const vector<T, 3> yaxis = core::cross(zaxis, xaxis);
+
+	matrix<T, 3, 4> r;
+	r[0] = vector<T, 3>(xaxis, -dot(xaxis, position));
```

Comment:
note that

- `(xaxis, -dot(xaxis, position)` -> `xaxis` is 3 dimension vector, your dot product is a scalar
- `r[i]` is `vector<T, 4>`

and you are 

- trying to call constructor which is reserved for 4 dimension vectors with `vector<T, 3>` which doesn't exist hence this code won't compile, there is no implicit conversion to do it
- assign the result to `r[i]`

Reply:
fixed